File: sdk/go/otel.go
// sdk/go/otel.go
// ─────────────────────────────────────────────────────────────
// OpenTelemetry setup for Go services.
// Initializes traces, metrics, and structured logging via the
// OTel SDK with OTLP gRPC export to the collector.
//
// INSTALL:
//   go get go.opentelemetry.io/otel \
//     go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc \
//     go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc \
//     go.opentelemetry.io/otel/sdk/trace \
//     go.opentelemetry.io/otel/sdk/metric \
//     go.opentelemetry.io/otel/sdk/resource \
//     go.opentelemetry.io/semconv/v1.26.0 \
//     go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp \
//     go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc
//
// USAGE:
//   shutdown, err := otelsetup.Init(context.Background())
//   if err != nil { log.Fatal(err) }
//   defer shutdown(context.Background())
//
//   // Wrap your HTTP handler:
//   http.Handle("/api/", otelhttp.NewHandler(myHandler, "api"))
//
// ENV VARS:
//   OTEL_SERVICE_NAME              my-go-service
//   OTEL_EXPORTER_OTLP_ENDPOINT   http://otel-collector:4317
//   OTEL_DEPLOYMENT_ENVIRONMENT   production
//   SENTRY_DSN                     http://KEY@errors.yourdomain.com/1
// ─────────────────────────────────────────────────────────────

package otelsetup

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

// ShutdownFunc flushes and closes all OTel providers.
// Defer this in main() after calling Init().
type ShutdownFunc func(ctx context.Context) error

// Init sets up OTel trace and metric providers.
// Returns a shutdown function to call on graceful exit.
//
// Example (main.go):
//
//	func main() {
//	    ctx := context.Background()
//	    shutdown, err := otelsetup.Init(ctx)
//	    if err != nil { log.Fatal(err) }
//	    defer func() {
//	        if err := shutdown(ctx); err != nil {
//	            log.Printf("OTel shutdown error: %v", err)
//	        }
//	    }()
//	    // ... start your server
//	}
func Init(ctx context.Context) (ShutdownFunc, error) {
	endpoint := getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "otel-collector:4317")
	// Strip any scheme prefix — gRPC uses bare host:port
	endpoint = stripScheme(endpoint)

	serviceName    := getenv("OTEL_SERVICE_NAME",             "go-service")
	serviceVersion := getenv("SERVICE_VERSION",               "0.0.0")
	environment    := getenv("OTEL_DEPLOYMENT_ENVIRONMENT",   "production")

	// ── Resource ───────────────────────────────────────────────
	res, err := resource.Merge(
		resource.Default(),
		resource.NewWithAttributes(
			semconv.SchemaURL,
			semconv.ServiceName(serviceName),
			semconv.ServiceVersion(serviceVersion),
			semconv.DeploymentEnvironment(environment),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("otel: build resource: %w", err)
	}

	// ── gRPC connection ────────────────────────────────────────
	conn, err := grpc.NewClient(
		endpoint,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		return nil, fmt.Errorf("otel: gRPC dial %s: %w", endpoint, err)
	}

	// ── Trace provider ─────────────────────────────────────────
	traceExporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
	if err != nil {
		return nil, fmt.Errorf("otel: trace exporter: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithResource(res),
		sdktrace.WithBatcher(traceExporter,
			sdktrace.WithBatchTimeout(time.Second),
		),
		// Adjust sample rate via OTEL_TRACES_SAMPLER_ARG env var in the collector
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
	)
	otel.SetTracerProvider(tp)

	// ── Metric provider ────────────────────────────────────────
	metricExporter, err := otlpmetricgrpc.New(ctx, otlpmetricgrpc.WithGRPCConn(conn))
	if err != nil {
		return nil, fmt.Errorf("otel: metric exporter: %w", err)
	}

	mp := sdkmetric.NewMeterProvider(
		sdkmetric.WithResource(res),
		sdkmetric.WithReader(
			sdkmetric.NewPeriodicReader(metricExporter,
				sdkmetric.WithInterval(30*time.Second),
			),
		),
	)
	otel.SetMeterProvider(mp)

	// ── Propagation (W3C TraceContext + Baggage) ───────────────
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	log.Printf("[OTel] ✓ %s@%s → %s | env: %s", serviceName, serviceVersion, endpoint, environment)

	// ── GlitchTip (Sentry-compatible) ─────────────────────────
	initSentry()

	// ── Shutdown function ──────────────────────────────────────
	return func(ctx context.Context) error {
		var errs []error
		if err := tp.Shutdown(ctx); err != nil {
			errs = append(errs, err)
		}
		if err := mp.Shutdown(ctx); err != nil {
			errs = append(errs, err)
		}
		if err := conn.Close(); err != nil {
			errs = append(errs, err)
		}
		if len(errs) > 0 {
			return fmt.Errorf("otel shutdown errors: %v", errs)
		}
		return nil
	}, nil
}

// Tracer returns a named tracer from the global provider.
// Use this to create spans in your application code.
//
//	tracer := otelsetup.Tracer("payments")
//	ctx, span := tracer.Start(ctx, "process-payment")
//	defer span.End()
func Tracer(name string) interface{ Start(ctx context.Context, spanName string, opts ...interface{}) (context.Context, interface{}) } {
	return otel.Tracer(name)
}

// Meter returns a named meter from the global provider.
// Use this to create custom counters, histograms, and gauges.
//
//	meter := otelsetup.Meter("api")
//	reqCount, _ := meter.Int64Counter("api.requests.total")
//	reqCount.Add(ctx, 1, metric.WithAttributes(attribute.String("route", "/users")))
func Meter(name string) interface{} {
	return otel.Meter(name)
}

func initSentry() {
	// Install: go get github.com/getsentry/sentry-go
	//
	// dsn := os.Getenv("SENTRY_DSN")
	// if dsn == "" { return }
	// if err := sentry.Init(sentry.ClientOptions{
	//     Dsn:              dsn,
	//     Environment:      os.Getenv("OTEL_DEPLOYMENT_ENVIRONMENT"),
	//     TracesSampleRate: 0, // tracing handled by OTel
	// }); err != nil {
	//     log.Printf("[GlitchTip] init error: %v", err)
	//     return
	// }
	// defer sentry.Flush(2 * time.Second)
	// log.Println("[GlitchTip] ✓ error tracking initialized")
}

func getenv(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func stripScheme(endpoint string) string {
	for _, prefix := range []string{"http://", "https://", "grpc://"} {
		if len(endpoint) > len(prefix) && endpoint[:len(prefix)] == prefix {
			return endpoint[len(prefix):]
		}
	}
	return endpoint
}

File: sdk/rust/otel.rs
// sdk/rust/otel.rs
// ─────────────────────────────────────────────────────────────
// OpenTelemetry setup for Rust services using Tokio async runtime.
// Integrates with the `tracing` ecosystem for structured logging.
//
// CARGO.TOML DEPENDENCIES:
//   [dependencies]
//   opentelemetry             = { version = "0.22", features = ["trace", "metrics"] }
//   opentelemetry-otlp        = { version = "0.15", features = ["grpc-tonic", "metrics"] }
//   opentelemetry_sdk         = { version = "0.22", features = ["rt-tokio"] }
//   opentelemetry-semantic-conventions = "0.14"
//   tracing                   = "0.1"
//   tracing-opentelemetry     = "0.23"
//   tracing-subscriber        = { version = "0.3", features = ["env-filter", "json"] }
//   tokio                     = { version = "1", features = ["full"] }
//   tonic                     = "0.11"
//   sentry                    = { version = "0.32", features = ["tracing"] }  # optional
//
// USAGE:
//   #[tokio::main]
//   async fn main() {
//       let _guard = otel::init().expect("OTel init failed");
//       // your app here
//       // _guard dropped on exit → flushes telemetry
//   }
//
// ENV VARS:
//   OTEL_SERVICE_NAME              my-rust-service
//   OTEL_EXPORTER_OTLP_ENDPOINT   http://otel-collector:4317
//   OTEL_DEPLOYMENT_ENVIRONMENT   production
//   SENTRY_DSN                     http://KEY@errors.yourdomain.com/1
//   RUST_LOG                       info   (controls tracing subscriber level)
// ─────────────────────────────────────────────────────────────

use opentelemetry::{global, KeyValue};
use opentelemetry_otlp::{ExportConfig, WithExportConfig};
use opentelemetry_sdk::{
    metrics::{MeterProviderBuilder, PeriodicReader},
    runtime,
    trace::{BatchConfig, RandomIdGenerator, Sampler, TracerProvider},
    Resource,
};
use opentelemetry_semantic_conventions::{
    resource::{DEPLOYMENT_ENVIRONMENT, SERVICE_NAME, SERVICE_VERSION},
    SCHEMA_URL,
};
use std::env;
use tracing_opentelemetry::OpenTelemetryLayer;
use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt, EnvFilter};

/// Guard that shuts down OTel providers on drop.
/// Keep this alive for the lifetime of your application.
///
/// ```rust
/// #[tokio::main]
/// async fn main() {
///     let _guard = otel::init().expect("OTel init failed");
///     // your application runs here
///     // _guard is dropped at end of main → flushes all telemetry
/// }
/// ```
pub struct OtelGuard {
    tracer_provider: TracerProvider,
}

impl Drop for OtelGuard {
    fn drop(&mut self) {
        if let Err(e) = self.tracer_provider.shutdown() {
            eprintln!("[OTel] Tracer shutdown error: {e}");
        }
        global::shutdown_tracer_provider();
    }
}

/// Initialize OpenTelemetry traces, metrics, and structured logging.
/// Returns an OtelGuard — keep it alive until your process exits.
pub fn init() -> Result<OtelGuard, Box<dyn std::error::Error + Send + Sync>> {
    let service_name    = env_or("OTEL_SERVICE_NAME",             "rust-service");
    let service_version = env_or("SERVICE_VERSION",               "0.0.0");
    let environment     = env_or("OTEL_DEPLOYMENT_ENVIRONMENT",   "production");
    let endpoint        = env_or("OTEL_EXPORTER_OTLP_ENDPOINT",  "http://otel-collector:4317");

    // ── Shared resource ───────────────────────────────────────
    let resource = Resource::from_schema_url(
        [
            KeyValue::new(SERVICE_NAME,             service_name.clone()),
            KeyValue::new(SERVICE_VERSION,          service_version),
            KeyValue::new(DEPLOYMENT_ENVIRONMENT,   environment.clone()),
        ],
        SCHEMA_URL,
    );

    // ── Tracer provider ───────────────────────────────────────
    let export_config = ExportConfig {
        endpoint: endpoint.clone(),
        ..Default::default()
    };

    let tracer_provider = opentelemetry_otlp::new_pipeline()
        .tracing()
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_export_config(export_config.clone()),
        )
        .with_trace_config(
            opentelemetry_sdk::trace::Config::default()
                .with_sampler(Sampler::AlwaysOn)
                .with_id_generator(RandomIdGenerator::default())
                .with_resource(resource.clone()),
        )
        .with_batch_config(BatchConfig::default())
        .install_batch(runtime::Tokio)?;

    global::set_tracer_provider(tracer_provider.clone());

    // ── Metric provider ───────────────────────────────────────
    let metric_exporter = opentelemetry_otlp::new_exporter()
        .tonic()
        .with_export_config(export_config)
        .build_metrics_exporter(
            Box::new(opentelemetry_sdk::metrics::reader::DefaultTemporalitySelector::new()),
            Box::new(opentelemetry_sdk::metrics::reader::DefaultAggregationSelector::new()),
        )?;

    let meter_provider = MeterProviderBuilder::default()
        .with_resource(resource)
        .with_reader(
            PeriodicReader::builder(metric_exporter, runtime::Tokio)
                .with_interval(std::time::Duration::from_secs(30))
                .build(),
        )
        .build();

    global::set_meter_provider(meter_provider);

    // ── Tracing subscriber (structured logging + OTel spans) ──
    // Reads log level from RUST_LOG env var (e.g. RUST_LOG=info,my_crate=debug)
    let tracer = global::tracer(service_name.clone());
    let otel_layer = OpenTelemetryLayer::new(tracer);

    tracing_subscriber::registry()
        .with(EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info")))
        .with(
            // JSON structured logs for Loki ingestion
            tracing_subscriber::fmt::layer().json(),
        )
        .with(otel_layer)
        .init();

    // ── GlitchTip / Sentry ────────────────────────────────────
    // Uncomment after adding `sentry = "0.32"` to Cargo.toml:
    //
    // let _sentry_guard = sentry::init((
    //     std::env::var("SENTRY_DSN").unwrap_or_default(),
    //     sentry::ClientOptions {
    //         environment: Some(environment.into()),
    //         traces_sample_rate: 0.0, // tracing handled by OTel
    //         ..Default::default()
    //     },
    // ));

    tracing::info!(
        service = %service_name,
        endpoint = %endpoint,
        "[OTel] ✓ instrumentation initialized"
    );

    Ok(OtelGuard { tracer_provider })
}

/// Convenience: get a named meter for custom business metrics.
///
/// ```rust
/// let meter = otel::meter("payments");
/// let counter = meter.u64_counter("payments.total").init();
/// counter.add(1, &[KeyValue::new("currency", "NGN")]);
/// ```
pub fn meter(name: &'static str) -> opentelemetry::metrics::Meter {
    global::meter(name)
}

fn env_or(key: &str, default: &str) -> String {
    env::var(key).unwrap_or_else(|_| default.to_owned())
}

// ─────────────────────────────────────────────────────────────
// AXUM INTEGRATION EXAMPLE
// ─────────────────────────────────────────────────────────────
//
// Add to Cargo.toml:
//   tower-http = { version = "0.5", features = ["trace"] }
//
// In main.rs:
//
// use tower_http::trace::TraceLayer;
// use axum::Router;
//
// let app = Router::new()
//     .route("/api/users", get(list_users))
//     .layer(TraceLayer::new_for_http());
//
// ─────────────────────────────────────────────────────────────
// ACTIX-WEB INTEGRATION EXAMPLE
// ─────────────────────────────────────────────────────────────
//
// Add to Cargo.toml:
//   actix-web-opentelemetry = "0.18"
//
// In main.rs:
//
// use actix_web_opentelemetry::RequestTracing;
//
// App::new()
//     .wrap(RequestTracing::new())
//     .service(...)
// ─────────────────────────────────────────────────────────────
